keps/sig-network/20191020-default-deny-pod-traffic

```diff
@@ -0,0 +1,270 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Make pod-to-pod communication in Kubernetes secure by default.
+Pods should require an explicit NetworkPolicy ingress configuration,
+in order to receive traffic.
+
+This proposal explicitly targets ingress traffic.
+Egress traffic is not _currently_ being targeted,
+not is not yet deemed out of scope.
+
+## Motivation
+
+Kubernetes and Kubernetes workloads are notoriously easy to leave insecure.
+Many users are not aware of relevant settings (like NetworkPolicy),
+or choose not to configure them due to friction and learning curves.
+This is a systemic pattern,
+NetworkPolicy is one example.
+
+TODO pull examples
+
+### Goals
+
+* Restrict pod ingress by default on a cluster.
+
+### Non-Goals
+
+What is out of scope for this KEP?
+Listing non-goals helps to focus discussion and make progress.
+
+## Proposal
+
+There are multiple options for design/implementation.
+
+Restricting network access requires a CNI driver that supports NetworkPolict.
+All options will require codifying NetworkPolicy support as a requirement,
+or creating a clear user experience when using a CNI driver that does not support NetworkPolicy.
+
+### Option 1: Change default NetworkPolicy behavior
+
+The most conceptually simple option is to change default NetworkPolicy behavior.
+Currently,
+NetworkPolicy allows all traffic to pods that are not selected by a NetworkPolicy. [1]
+
+If all pods are treated as isolated,
+users must explicitly create NetworkPolicy objects
+(either as needed, or by allowing all ingress and _explicitly_ enabling the existing behavior).
+This would potentially require a v2 API, due to changes in behavior (if not the API itself).
+
+### Option 2: Inject explicit deny-all NetworkPolicy objects
+
+If making changes to NetworkPolicy API is too undesirable,
+we could automatically apply NetworkPolicy objects to achieve the same effect.
+
+There is no precedent(?) or mechanism(?) for performing pre-creation actions
+(IE the opposite of finalizers).
+
+A simple controller could monitor for new namespaces,
+but it would be subject to race conditions,
+where new pods would be subject to the default-open behavior before the NetworkPolicy is created.
+
```

Comment:
IMHO Option 3 would be to implement `GlobalNetworkPolicy` similar to what Calico and Clilum already supports and would leave the freedom to operators to decide if they want the default deny rule or not (globally).

Reply:
As someone who has had to write a custom controller (to create allow to dns, allow to web-proxy, allow to monitoring rules and deny everything in prod but allow all in other environments rules for every namespace) because we did not want to use a "third party" globalNetworkPolicy implementation, this is my personal preference.

Reply:
IMHO current network policy is written from application owner view because only they know what traffic is expected for their application and what is not. If we want to enforce anything else from a cluster administrator view, we will want to have some new policy that is cluster-wide and serve as a guardrail that is not easily broken through by app owners. Delegating the policy completely to app owners doesn't seem to be very secure for the whole system.